File: src/fi/tuska/jalkametri/gui/NamedIconAdapter.java
/**
 * Copyright 2006-2011 Tuukka Haapasalo
 * 
 * This file is part of jAlkaMetri.
 * 
 * jAlkaMetri is free software: you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or (at your option)
 * any later version.
 * 
 * jAlkaMetri is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
 * more details.
 * 
 * You should have received a copy of the GNU General Public License along
 * with jAlkaMetri (LICENSE.txt). If not, see <http://www.gnu.org/licenses/>.
 */
package fi.tuska.jalkametri.gui;

import java.util.List;

import android.content.Context;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import fi.tuska.jalkametri.dao.NamedIcon;
import fi.tuska.jalkametri.util.Converter;

public class NamedIconAdapter<T extends NamedIcon> extends BaseAdapter {

    protected final Context context;
    protected final int defaultIconRes;
    private final List<T> icons;
    private final boolean vertical;
    private final Converter<NamedIcon, String> textFormatter;

    public NamedIconAdapter(Context c, List<T> icons, boolean vertical, int defaultIconRes) {
        this.context = c;
        this.icons = icons;
        this.vertical = vertical;
        this.textFormatter = null;
        this.defaultIconRes = defaultIconRes;
    }

    public NamedIconAdapter(Context c, List<T> icons, boolean vertical,
        Converter<NamedIcon, String> textFormatter, int defaultIconRes) {
        this.context = c;
        this.icons = icons;
        this.vertical = vertical;
        this.textFormatter = textFormatter;
        this.defaultIconRes = defaultIconRes;
    }

    @Override
    public int getCount() {
        return icons.size();
    }

    @Override
    public T getItem(int position) {
        return icons.get(position);
    }

    @Override
    public long getItemId(int position) {
        return 0;
    }

    // create a new TextIconView for each item referenced by the Adapter
    @Override
    public View getView(int position, View convertView, ViewGroup parent) {

        TextIconView view;
        if (convertView == null) {
            // If it's not recycled, create it
            view = new TextIconView(context, vertical);
        } else {
            view = (TextIconView) convertView;
        }

        T icon = getItem(position);

        // Set image
        int res = DrinkIconUtils.getDrinkIconRes(icon.getIcon());
        view.setImageResource(res != 0 ? res : defaultIconRes);
        // Set text
        String text = textFormatter != null ? textFormatter.convert(icon) : icon
            .getIconText(context.getResources());
        view.setText(text);
        return view;
    }

}

File: app/src/main/java/fi/tuska/jalkametri/gui/NamedIconAdapter.kt
package fi.tuska.jalkametri.gui

import android.content.Context
import android.view.Gravity.CENTER
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import fi.tuska.jalkametri.dao.NamedIcon
import fi.tuska.jalkametri.gui.DrinkIconUtils.getDrinkIconRes
import fi.tuska.jalkametri.util.Converter

open class NamedIconAdapter<out T : NamedIcon>(
        protected val context: Context,
        private val icons: List<T>,
        protected val defaultIconRes: Int,
        private val vertical: Boolean = true,
        private val gravity: Int = CENTER,
        private val textFormatter: Converter<NamedIcon, String>? = null) : BaseAdapter() {

    override fun getCount(): Int = icons.size

    override fun getItem(position: Int): T = icons[position]

    override fun getItemId(position: Int): Long = 0

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        val view = convertView as TextIconView? ?: TextIconView(context, vertical, gravity)
        val icon = getItem(position)

        val res = getDrinkIconRes(icon.icon)
        view.setImageResource(if (res != 0) res else defaultIconRes)

        view.text = textFormatter?.convert(icon) ?: icon.getIconText(context.resources)
        return view
    }

    fun inflater(): LayoutInflater =
            context.applicationContext.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater

}
